### ISSUE TITLE
[BUG]: Creation of py::tuple or py::list causes a crash with policy py::call_guard<py::gil_scoped_release>() on Ubuntu 22

### ISSUE THREAD

**Issue by ylegoc** [pybind/pybind11#4015]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [X] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

After some tests on my Ubuntu 22, I remarked that pybind11 was crashing at the creation of a py::tuple or py::list object in my project.

Calling the function test() defined below causes a segmentation fault.

Crashes on:
- Ubuntu 22 with pybind11 2.9.2 and also 2.4.3
- Ubuntu 22 with pybind11 2.9.1 from distrib package

No crash on:
- Ubuntu 20 with pybind11 2.9.2 or 2.4.3

By looking deeper into the problem, I found out:
Using the policy **py::call_guard&lt;py::gil_scoped_release&gt;()** causes the crash. Without the policy it works well.

I tried different compilation options that all resulted in a crash.

To summarize:
The creation of a py::tuple or py::list object causes a crash on my **Ubuntu 22** (gcc 11.2.0, Intel(R) Core(TM) i5-8500) with the policy **py::call_guard&lt;py::gil_scoped_release&gt;()**.

### Reproducible example code

```text
.def_static("test", [] {
    py::tuple tup = py::make_tuple(42, py::none(), "spam");
}, py::call_guard<py::gil_scoped_release>())
```

**Comment by ylegoc** [comment]
I removed the word regression from the title because this is a form of regression.
The simple code compiles and executes well on Ubuntu 20 (gcc 9.4).
However compiled on Ubuntu 22 (gcc 11.2), it crashes on Ubuntu 22.

That is why it seems to come from the gcc version but not from a bad commit.
I will try on master.

**Comment by ylegoc** [comment]
I run the test suite on my system and it succeeds. Then I modified test_pytypes.cpp:
`m.def("get_tuple", []() { return py::make_tuple(42, py::none(), "spam"); });`
`m.def("get_tuple_gil", []() { return py::make_tuple(42, py::none(), "spam"); }, py::call_guard<py::gil_scoped_release>());`

And test_pytypes.py:
`assert m.get_tuple() == (42, None, "spam")`
`assert m.get_tuple_gil() == (42, None, "spam")`

And it's crashing in the get_tuple_gil() call with a segmentation fault.

However maybe this is not a good practice and I should release the gil only for non pybind11 code?

**Comment by gst** [comment]
FWIW : I tried g++-11 + Ubuntu 20.04 (focal) and no segfault.

**Comment by henryiii** [comment]
You can’t modify the Python memory model with the GIL released. That’s why there’s a GIL. Making a tuple very much modifies Python memory. The crash is random and could be compiler dependent, it’s UB.
